File: 942. DI String Match.js
// 942. DI String Match
s = "IDID"
/**
 * @param {string} s
 * @return {number[]}
 */
var diStringMatch = function(s) {
    let source = Array.from({length: s.length + 1}, (_, i) => i);
    let res = [];
    for (let char of s) {
        char === 'I' ? res.push(source.shift()) : res.push(source.pop());
    }
    res.push(source.pop());
    return res;
};
console.log(diStringMatch(s))

File: 942. DI String Match.rb
#* 942. DI String Match
s = "IDID"
# @param {String} s
# @return {Integer[]}
def di_string_match(s)
    source = (0..s.size).to_a
    res = []
    s.chars.each do |char|
        char == 'I' ? res.push(source.shift) : res.push(source.pop)
    end
    res.push(source.pop)
    res
end
p di_string_match(s)